apollo-router/src/plugins/subgraph_auth.rs

```diff
@@ -0,0 +1,298 @@
+use std::collections::HashMap;
+use std::task::Context;
+use std::task::Poll;
+use std::time::SystemTime;
+
+use aws_sigv4::http_request::sign;
+use aws_sigv4::http_request::PayloadChecksumKind;
+use aws_sigv4::http_request::SignableBody;
+use aws_sigv4::http_request::SignableRequest;
+use aws_sigv4::http_request::SigningParams;
+use aws_sigv4::http_request::SigningSettings;
+use aws_types::Credentials;
+use schemars::JsonSchema;
+use serde::Deserialize;
+use tower::BoxError;
+use tower::Layer;
+use tower::ServiceBuilder;
+use tower::ServiceExt;
+use tower_service::Service;
+
+use crate::plugin::Plugin;
+use crate::plugin::PluginInit;
+use crate::register_plugin;
+use crate::services::subgraph;
+use crate::SubgraphRequest;
+
+register_plugin!("apollo", "subgraph_auth", SubgraphAuth);
+
+#[derive(Clone, JsonSchema, Deserialize)]
+#[serde(rename_all = "snake_case", deny_unknown_fields)]
+struct AWSSigV4Config {
+    access_key_id: String,
+    secret_access_key: String,
+    region: String,
+    service: String,
+}
+
+#[derive(Clone, JsonSchema, Deserialize)]
+#[serde(deny_unknown_fields)]
+enum AuthConfig {
+    #[serde(rename = "aws_sig_v4")]
+    AWSSigV4(AWSSigV4Config),
+}
+
+#[derive(Clone, JsonSchema, Deserialize)]
+#[serde(rename_all = "snake_case", deny_unknown_fields)]
+struct Config {
+    #[serde(default)]
+    all: Option<AuthConfig>,
+    #[serde(default)]
+    subgraphs: HashMap<String, AuthConfig>,
+}
+
+struct SubgraphAuth {
+    config: Config,
+}
+
+#[async_trait::async_trait]
+impl Plugin for SubgraphAuth {
+    type Config = Config;
+    async fn new(init: PluginInit<Self::Config>) -> Result<Self, BoxError> {
+        Ok(SubgraphAuth {
+            config: init.config,
+        })
+    }
+
+    fn subgraph_service(&self, name: &str, service: subgraph::BoxService) -> subgraph::BoxService {
+        let mut auth = self.config.all.as_ref();
+        if let Some(subgraph) = self.config.subgraphs.get(name) {
+            auth = Some(subgraph);
```

Comment:
it might be a good idea to support configuration merge and override like https://github.com/apollographql/router/blob/dev/apollo-router/src/plugins/traffic_shaping/mod.rs#L74-L95

So we could define global properties in `all`, like the region, and have other properties be per subgraph